### ISSUE TITLE
Cannot reproduce results

### ISSUE THREAD

**Issue by Haoye17** [llava-rlhf/LLaVA-RLHF#8]
I'm trying to reproduce the [LLaVA-RLHF-13b-v1.5-336](https://huggingface.co/zhiqings/LLaVA-RLHF-13b-v1.5-336) model results on POPE, but I got results **slightly lower** than your reports. In my reproduction, the average F1 score is 79.5, and the average "Yes" rate is 36.8. Are there any other things I should do to reproduce the exact results in your paper? Thanks a lot!

Here are my settings:

- **model-ckpt**: [LLaVA-RLHF-13b-v1.5-336](https://huggingface.co/zhiqings/LLaVA-RLHF-13b-v1.5-336)
- **inference code**: I follow the [model_vqa.py](https://github.com/haotian-liu/LLaVA/blob/main/llava/eval/model_vqa.py) code in the newest LLaVA repo:

I load the model checkpoint like this:
```
tokenizer, model, image_processor, context_len = load_pretrained_model(model_path, args.model_base, args.model_name)
if args.lora_path is not None:
    model = PeftModel.from_pretrained(
        model,
        args.lora_path,
    )
    model = model.merge_and_unload()
```
I also follow your instruction to add additional prompt after the question:
```
if args.add_yesno:
    # qs = line["question"] + " Please answer yes or no."
    qs = line["question"] + "\nAnswer the question using a single word or phrase."
else:
    qs = line["question"]
```

- **generation arguments**: temperature=0.2, num_beams=1, top-p=None, do_sample=True, max_new_tokens=1024

**Comment by sIncerass** [comment]
Hi @Haoye17, thanks for your interest in our work! two things that is used to reproduce our results are 
1. load in bf16 for lora components and general model:
```
        bits = 16
        dtype = torch.bfloat16
        compute_dtype = torch.bfloat16
        model = LlavaLlamaForCausalLM.from_pretrained(
            model_path,
            device_map={"": "cuda:0"},
            torch_dtype=dtype,
            load_in_4bit=(bits == 4),
            load_in_8bit=(bits == 8),
            quantization_config=BitsAndBytesConfig(
                load_in_4bit=(bits == 4),
                load_in_8bit=(bits == 8),
                llm_int8_threshold=6.0,
                llm_int8_skip_modules=["mm_projector", "lm_head"],
                llm_int8_has_fp16_weight=False,
                bnb_4bit_compute_dtype=compute_dtype,
                bnb_4bit_use_double_quant=True,
                bnb_4bit_quant_type="nf4",
            ),
        )
```
2. use padding for image preprocessing 
```
--image_aspect_ratio pad
```

Could you please try these to see if the results could be reproduced at your end?

**Comment by Haoye17** [comment]
Thanks! We can reproduce it now! 😊

**Comment by Haoye17** [comment]
Sorry for bothering you again! We recently found that although we can now reproduce the POPE results in your paper, we cannot reproduce the results on your MMHal-Bench😢. We use your evaluation code `model_vqa.py` to test on MMHal, and the modification and runing scripts are as follow:
1. We use the transformers==4.28.1 to run the code, so we change the model loading part:
```
# modification in model_vqa.py
model = LlavaLlamaForCausalLM.from_pretrained(
            model_path,
            device_map={"": "cuda:0"},
            torch_dtype=dtype,
            # load_in_4bit=(bits == 4),
            # load_in_8bit=(bits == 8),
            # quantization_config=BitsAndBytesConfig(
            #     load_in_4bit=(bits == 4),
            #     load_in_8bit=(bits == 8),
            #     llm_int8_threshold=6.0,
            #     llm_int8_skip_modules=["mm_projector", "lm_head"],
            #     llm_int8_has_fp16_weight=False,
            #     bnb_4bit_compute_dtype=compute_dtype,
            #     bnb_4bit_use_double_quant=True,
            #     bnb_4bit_quant_type="nf4",
            # ),
        )
```
2. Our running script:
```
python ./Eval/model_vqa.py \
--model-path ./models/llava-rlhf/sft_model \
--model-base llava-rlhf-13b-v1.5-336 \
--question-file response_template_addpath.json \
--answers-file malhal_answer__llava-rlhf-13b-v1.5-336_rm.jsonl \
--use-qlora True \
--qlora-path /home/zhanghaoye/models/llava-rlhf/rm_lora_adapter_model \
--test-prompt ''
```
We also test the lora weight in the rlhf_lora_adapter_model, but it still cannot work correctly.

Is there anything wrong with that? Thank you again!

**Comment by Edward-Sun** [comment]
Hi @Haoye17 , 
> --qlora-path /home/zhanghaoye/models/llava-rlhf/rm_lora_adapter_model
`rm_lora_adapter_model` is only used for the reward model. Could you try `rlhf_lora_adapter_model`?

**Comment by Haoye17** [comment]
Hi! As I mentioned above, we have tried both `rm_lora_adapter_model` and `rlhf_lora_adapter_model`, but they got similar results on MMHal-Bench, which is around 2.0 on the average score😢.

**Comment by Edward-Sun** [comment]
Sorry that I missed that information. Recently we updated some public code that is not consistent with our original codebase (including a modification to bf16 layer normalization in `llama`, which might affect the final performance). Could you create a new environment, install the `llava` and related package following the [guideline here](https://github.com/llava-rlhf/LLaVA-RLHF/tree/main/llava_setup)?

**Comment by Edward-Sun** [comment]
By the way, could you get a score around 2.4 for the LLaVA-SFT+-13b model? The results can be obtained by commenting out the lora module:

```bash
python ./Eval/model_vqa.py \
--model-path ./models/llava-rlhf/sft_model \
--model-base llava-sft-13b-v1.5-336 \
--question-file response_template_addpath.json \
--answers-file malhal_answer__llava-sft-13b-v1.5-336_rm.jsonl \
--use-qlora True \
--test-prompt ''
```

**Comment by Haoye17** [comment]
Hi! I've tried following the instructions, however, there is an error when I install the bitsandbytes==0.41.0, I'm going to try it again if there's nothing else that seems to be wrong. I'll also try the SFT+ model to see if it can work!

**Comment by Edward-Sun** [comment]
Since we didn't use quantization in QLoRA, `bitsandbytes==0.41.0` might not be necessary. Perhaps it's ok to use older version of bitsandbytes or just comment out the lines of code that import bitsandbytes

**Comment by Haoye17** [comment]
Hi! I've tried the environment in your instruction, but both the rlhf model and the sft+ model have lower score than yours. Which is around 2.3 for the rlhf model and 2.14 for the sft+ model. I've also tried to reproduce the POPE again using your original script in the environment, and it looks good to match your results. But when I tried to reproduce LLaVA-Bench using `Eval/eval_scripts/eval_llava_bench.sh`, the result was still worse than your report. I think there might be something wrong with my generation period🤔. By the way, I find that in `eval_llava_bench.sh`, you use `--short_eval True`, is it normal? And would you mind uploading the MMHal-Bench script you are using and the inference results for LLaVA-Bench and MMHal-Bench? Thanks a lot!

**Comment by sIncerass** [comment]
Thanks for pointing it out, we just corrected the ` eval_llava_bench.sh`.

**Comment by Edward-Sun** [comment]
Hi, Sorry for the long wait, and thank you for your patience.

We rerun our script for evaluating our model on the MMHal-Bench and found the performance to vary on a scale of 0.1 in different runs, potentially due to the non-deterministic generation of LMMs and the non-deterministic generation of the GPT-4-based judge scores. We have updated the code to `do_sample=False` when evaluating on MMHal-Bench and set `temperature=0.0` for `openai.ChatCompletion.create`, but still observe a variance of about 0.02, which is due to the non-deterministic nature of the GPT-4 API.

Our new (more deterministic) evaluation on the released 13b RLHF ckpt achieves MMHal-Bench scores of 2.43, 2.36, 2.38 in three runs. We also uploaded a [new checkpoint](https://huggingface.co/zhiqings/LLaVA-RLHF-13b-v1.5-336/tree/main/rlhf_lora_adapter_model_ckpt250) (trained for 250 PPO steps, while the previously released one is trained for 240 PPO steps), which achieves an MMBench score of 59.9 and achieves MMHal-Bench scores of 2.56, 2.58, 2.47, which may better match the score in the paper.

We are also actively looking into prompt engineering techniques to make GPT-4 judge scores more stable, like predicting the score first and then generating rationals.

For now, we recommend running the GPT-4-based evaluation several times and taking the average score. We will update the results in the table accordingly, and we welcome your feedback.
